DataDog/datadog-agent
# [CSM] `imds` event type

### What does this PR do?



This PR introduces the `imds` event. This event is generated by capturing the raw HTTP traffic to and from the IP of the cloud metadata services (i.e. `169.254.169.254`). For now, only AWS credentials are supported in response, and we're storing only the AccessKeyID (secrets and other sensitive credentials aren't even parsed).

New SecL rules can be written to be notified when specific IMDS traffic patterns are detected (such as the use of IMDSv1 instead of IMDSv2 on AWS). For example:
```
imds.cloud_provider == "aws" && imds.aws.is_imds_v2 == false
```

`imds` events are also added to Activity Trees, meaning they can be used to monitor continuously which processes are making `imds` requests, along with retrieving which processes resolved which `AccessKeyID`. This `AccessKeyID` can then be leveraged to correlate signals in the backend. In more technical terms, `imds` events are sent as part of Activity Dumps and Drift events.

### Motivation



There are 2 main motivation for this change:
- help monitor the usage of IMDSv1 in order to make a transition to IMDSv2 easier (and make sure it won't break services)
- keep track of which processes in an infrastructure had access to which `AccessKeyID` so that process, service and app context can be provided to Cloud based signals. It will also allow for accurate correlation between signals from different products.

### Describe how to test/QA your changes



In order to test this PR you'll have to setup a local IMDS server and generate traffic to it.

1) Set up a local dummy interface which will be used to route traffic to the instance metadata service IP address `169.254.169.254`. To do so, run the following commands:

```
ip link add dummy0 type dummy
ip addr add 169.254.169.254/32 dev dummy0
ip link set dummy0 up
```

2) Create a dummy IMDS server that binds on the IMDS IP address. A go example is provided below:

```golang
package main

package main

import (
	"encoding/json"
	"fmt"
	"net/http"
)

func main() {
	http.HandleFunc("/latest/meta-data/iam/security-credentials/hello", func(w http.ResponseWriter, r *http.Request) {
		// Define your custom JSON data
		data := map[string]interface{}{
			"AccessKeyId":     "ASIAIOSFODNN7EXAMPLE",
			"SecretAccessKey": "FQoDYXdzEL3/FQoDYXdzEL3/FQoDYXdzEL3",
			"Token":           "FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3FQoDYXdzEL3",
			"Expiration":      "2324-05-01T12:00:00Z",
			"Code":            "Success",
			"LastUpdated":     "2012-04-26T16:39:16Z",
			"Type":            "AWS-HMAC",
		}

		// Convert data to JSON
		response, err := json.Marshal(data)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

		// Set Content-Type header to application/json
		w.Header().Set("Content-Type", "application/json")

		// Write JSON response
		w.Write(response)
	})
	fmt.Println("Server listening on port 8080...")
	http.ListenAndServe("169.254.169.254:8080", nil)
}

```

3) Add the following policy to your policy folder:

```
version: '1'
rules:
- id: imds_v1_rule
  version: 1
  description: Detect IMDSv1 traffic
  expression: imds.cloud_provider == "aws" && imds.aws.is_imds_v2 == false
  agent_version: ''
  filters: []
- id: touch_tmp_test
  version: 1
  description: touch /tmp/test
  expression: open.file.path == "/tmp/test"
```

4) Start the agent.

5) There are 3 things to test:

- Run the following command and check if an `imds_v1_rule` event was generated:

```bash
curl http://169.254.169.254:8080/latest/meta-data/iam/security-credentials/hello
```

- Run the following command and check if a populated `aws_security_credentials` field with the access key id `ASIAIOSFODNN7EXAMPLE` was sent as part of the process context of the `touch_tmp_test` event:

```bash
python3 -c 'import requests; requests.get("http://169.254.169.254:8080/latest/meta-data/iam/security-credentials/hello"); open("/tmp/test", "w")'
```

- Do the same thing within a container and make sure:
  - that IMDS nodes are properly created and serialized (including in the `.dot` graph)
  - that drift events are generated for IMDS events
